docs/architecture/architecture.md

```diff
@@ -0,0 +1,671 @@
+about completely deleting the whole thing: the pod, its containers, all `cc-shim`
+instances and the virtual machine itself.
+Also, the latter case is only supported by Kubernetes whereas Docker always asks
+for the former one.
+
+Below we will focus on describing the entire pod deletion process and we will
+look at the generic, multi-containers pod setup. Docker is just a particular case
+for single container pods.
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let
+  it know on which pod the container it is trying to to `delete` is running.
+2. `cc-runtime` terminates all container processes within a pod by sending the
+  `KILLCONTAINER` command to all of them.
+3. As each container process termination will be forwarded to all associated `cc-shim`
+  instances, `cc-runtime` waits for all of them to terminate as well.
+4. `cc-runtime` sends the `UnregisterVM` command to `cc-proxy`, to let it know
+  that the virtual machine that used to host the pod should no longer be used.
+5. `cc-runtime` explicitly shuts the virtual machine down.
+6. The host namespaces are cleaned up and destroyed. In particular, `cc-runtime`
+  offloads the networking namespace cleanup path by calling into the specific
+  networking model (CNM or CNI) removal method.
+7. All remaining pod related resources on the host are deleted.
+
+## Proxy
+
+`cc-proxy` is a daemon offering access to the VM [`cc-agent`](https://github.com/clearcontainers/agent)
+to multiple `cc-shim` and `cc-runtime` clients. Only a single instance of `cc-proxy`
+per host is necessary as it can be used for several different VMs.
+Its main role is to:
+- Arbitrate access to the `cc-agent` control channel between all the `cc-runtime`
+  instances and the `cc-shim` ones.
+- Route the I/O streams and signals between the various `cc-shim` instances and
+  the `cc-agent`.
+
+`cc-proxy` API is available through a single socket for all `cc-shim` and `cc-runtime`
+instances to connect to. `cc-proxy` can be configured to use a UNIX or a TCP
+socket, and by default will handle connection over a UNIX socket.
+
+The protocol on the `cc-proxy` socket supports the following commands:
+- `RegisterVM`: Used by `cc-runtime` to let `cc-proxy` know about a newly created
+  virtual machine. When registering a new VM, `cc-runtime` will ask `cc-proxy`
+  for a token and pass that token to the `cc-shim` instance it will eventually
+  create for handling the container process inside that VM.
+- `UnregisterVM`: Does the opposite of what `RegisterVM` does, indicating to
+  the proxy it should release resources created by `RegisterVM`.
+- `AttachVM`: It can be used to associate clients to an already known VM.
+  Optionally `AttachVM` senders can ask `cc-proxy` for a token. In the OCI `exec`
+  command case, this token will be used by the `cc-shim` instance that monitors
+  the executed process inside the container.
+- `Hyper` : This payload will forward an hyperstart command to the `cc-agent`.
```

Comment:
extraneous space ' ' after `Hyper`, before the :

Reply:
ack